cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,152 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+            for package in packages
+            if "file:" not in package.resolved_url
+        }
+
+
+def extract_git_info_npm(vcs_url: str) -> Dict[str, str]:
+    """
+    Extract important info from a VCS requirement URL.
+
+    Given a URL such as git+ssh://user@host/namespace/repo.git#9e164b970
+
+    this function will extract:
+    - the "clean" URL: ssh://user@host/namespace/repo.git
+    - the git ref: 9e164b970
+
+    The clean URL and ref can be passed straight to scm.Git to fetch the repo.
+    The host, namespace and repo will be used to construct the file path under deps/npm.
+
+    :param vcs_url: The URL of a VCS requirement, must be valid (have git ref in path)
+    :return: Dict with url, ref, host, namespace and repo keys
+    """
+    # If scheme is git+protocol://, keep only protocol://
+    # Do this before parsing URL, otherwise urllib may not extract URL params
+    if vcs_url.startswith("git+"):
+        vcs_url = vcs_url[len("git+") :]
+
+    ref = vcs_url[-40:]  # Take the last 40 characters (the git ref)
+    clean_url = vcs_url[:-41]  # Drop the last 41 characters ('#' + git ref)
+
+    url = urlparse(clean_url)
+    namespace_repo = url.path.strip("/")
+    if namespace_repo.endswith(".git"):
+        namespace_repo = namespace_repo[: -len(".git")]
+
+    # Everything up to the last '/' is namespace, the rest is repo
+    namespace, _, repo = namespace_repo.rpartition("/")
+
+    return {
+        "url": clean_url,
+        "ref": ref.lower(),
+        "host": str(url.hostname),
+        "namespace": namespace,
+        "repo": repo,
+    }
+
+
+def clone_repo_pack_archive(
+    vcs: str,
+    download_dir: RootedPath,
+) -> RootedPath:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param url: URL for file download
+    :param download_dir: Output folder where dependencies will be downloaded
+    :raise FetchError: If download failed
+    """
+    info = extract_git_info_npm(vcs)
+    download_path = RootedPath(
+        Path(
+            download_dir,
+            info["host"],  # host
+            info["namespace"],
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def get_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of dependencies with downloaded paths
+    """
+    files_to_download = {}
```

Comment:
```suggestion
    files_to_download: dict[str, Any] = {}
```

^ this would also be needed for my suggestions to type-check

Notice that introducing untyped JSON-like object in a typed context always causes issues. Using properly typed objects makes your life easier.

Reply:
(but for now this is fine)

For future reference, it would probably be easier to work with something like this

```python
files_to_download: list[tuple[str, RootedPath, ChecksumInfo]] = []
...
async_download_files({url: path for url, path, _ in files_to_download})
```
One level above would that would be a [NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) (probably overkill in this case)

```python
class FileToDownload(NamedTuple):
    url: str
    download_path: str | PathLike[str]
    checksum: ChecksumInfo

files_to_download: list[FileToDownload] = []
```

Reply:
this is th s*it I was missing yesterday :( sigh
I love you <3

Reply:
yesterday I was thinking of rewriting it to some dataclass but I decided till morning with decision, Thanks Adam